lib/semantics/check-declarations.cc

```diff
@@ -457,6 +469,137 @@ static bool ConflictsWithIntrinsicAssignment(const Procedure &proc) {
+  std::optional<parser::MessageFixedText> msg;
+  if (!proc.functionResult.has_value()) {
+    msg = "%s procedure '%s' must be a function"_err_en_US;
+  } else if (proc.functionResult->IsAssumedLengthCharacter()) {
+    msg = "%s function '%s' may not have assumed-length CHARACTER(*)"
+          " result"_err_en_US;
+  } else if (!CheckNumberOfArgs(msg, kind, proc.dummyArguments.size())) {
+    // error is in msg
+  } else if (!CheckDefinedOperatorArg(opName, specific, proc, 0) |
+      !CheckDefinedOperatorArg(opName, specific, proc, 1)) {
+    return false;  // error was reported
+  } else if (ConflictsWithIntrinsicOperator(kind, proc)) {
+    msg = "%s function '%s' conflicts with intrinsic operator"_err_en_US;
+  } else {
+    return true;  // OK
+  }
+  SayWithDeclaration(specific, std::move(msg.value()),
+      parser::ToUpperCaseLetters(opName.ToString()), specific.name());
+  return false;
+}
+
+// If the number of arguments is wrong for this intrinsic operator, return
+// false and return the error message in msg.
+bool CheckHelper::CheckNumberOfArgs(
+    std::optional<parser::MessageFixedText> &msg, const GenericKind &kind,
+    std::size_t nargs) {
+  std::size_t min{2}, max{2};  // allowed number of args; default is binary
+  std::visit(
+      common::visitors{
+          [&](const common::NumericOperator &x) {
+            if (x == common::NumericOperator::Add ||
+                x == common::NumericOperator::Subtract) {
+              min = 1;  // + and - are unary or binary
+            }
+          },
+          [&](const common::LogicalOperator &x) {
+            if (x == common::LogicalOperator::Not) {
+              min = 1;  // .NOT. is unary
+              max = 1;
+            }
+          },
+          [](const common::RelationalOperator &) {
+            // all are binary
+          },
+          [](const GenericKind::OtherKind &x) {
+            CHECK(x == GenericKind::OtherKind::Concat);
+          },
+          [](const auto &) { DIE("expected intrinsic operator"); },
+      },
+      kind.u);
+  if (nargs >= min && nargs <= max) {
+    return true;
+  }
+  if (max == 1) {
+    msg = "%s function '%s' must have one dummy argument"_err_en_US;
+  } else if (min == 2) {
+    msg = "%s function '%s' must have two dummy arguments"_err_en_US;
+  } else {
+    msg = "%s function '%s' must have one or two dummy arguments"_err_en_US;
+  }
+  return false;
+}
+
+bool CheckHelper::CheckDefinedOperatorArg(const SourceName &opName,
+    const Symbol &symbol, const Procedure &proc, std::size_t pos) {
+  if (pos >= proc.dummyArguments.size()) {
+    return true;
+  }
+  auto &arg{proc.dummyArguments.at(pos)};
+  std::optional<parser::MessageFixedText> msg;
+  if (arg.IsOptional()) {
+    msg = "In %s function '%s', dummy argument '%s' may not be"
+          " OPTIONAL"_err_en_US;
+  } else if (const auto *dataObject{std::get_if<DummyDataObject>(&arg.u)};
```

Comment:
If you transpose the then & else parts here, it would avoid the unusual `nullptr` test in the predicate.

Reply:
It is unusual, but it makes the structure of the `if`s simpler. This way it's a series of: check for error and set error message. Changing it as you suggest requires nesting checks inside of this one.

If you feel strongly about avoiding this form of `if` I'm happy to change it. Otherwise I'll leave it as-is.